packages/jsii-pacmak/lib/targets/go/types/go-type-reference.ts

```diff
@@ -74,66 +75,109 @@ export class GoTypeRef {
    * Return the name of a type for reference from the `Package` passed in
    */
   public scopedName(scope: Package): string {
-    return this.scopedTypeName(scope, false);
+    return this.scopedTypeName(this.typeMap, scope, false);
   }
 
   public scopedInterfaceName(scope: Package): string {
-    return this.scopedTypeName(scope, true);
+    return this.scopedTypeName(this.typeMap, scope, true);
   }
 
   public scopedReferenceName(scope: Package): string {
-    return this.scopedTypeName(scope, false, true);
+    return this.scopedTypeName(this.typeMap, scope, false, true);
   }
 
-  private scopedTypeName(
-    scope: Package,
-    asInterface: boolean,
-    asRef = false,
-  ): string {
-    if (this.primitiveType) {
-      return this.primitiveType;
+  public get typeMap(): TypeMap {
+    return this.buildTypeMap(this);
+  }
+
+  private buildTypeMap(ref: GoTypeRef): TypeMap {
+    if (ref.primitiveType) {
+      return { primitive: ref.primitiveType };
+    } else if (ref.reference.arrayOfType) {
+      return { array: new GoTypeRef(this.root, ref.reference.arrayOfType) };
+    } else if (ref.reference.mapOfType) {
+      return { map: new GoTypeRef(this.root, ref.reference.mapOfType) };
+    } else if (ref.reference.unionOfTypes) {
+      return {
+        union: ref.reference.unionOfTypes.map(
+          (typeRef) => new GoTypeRef(this.root, typeRef),
+        ),
+      };
+    } else if (ref.reference.void) {
+      return { void: true };
     }
 
-    // type is an array
-    if (this.reference.arrayOfType) {
-      const innerName =
-        new GoTypeRef(this.root, this.reference.arrayOfType).scopedTypeName(
-          scope,
-          asInterface,
-          asRef,
-        ) ?? 'interface{}';
+    return { interface: ref };
+  }
+
+  /**
+   * Builds a map of interface to concrete types. This is passed to the runtime
+   * so that return types of interfaces can correctly be constructucted.
+   */
+  public scopedImplMap(scope: Package) {
+    const nextRef =
+      this.typeMap.array ?? this.typeMap.map ?? this.typeMap.interface;
 
-      return `[]${innerName}`;
+    if (nextRef) {
+      return [nextRef.scopedInterfaceName(scope), nextRef.scopedName(scope)];
     }
 
-    // type is a map
-    if (this.reference.mapOfType) {
+    return [];
+  }
+
+  private scopedTypeName(
+    typeMap: TypeMap,
+    scope: Package,
+    asInterface: boolean,
+    asRef = false,
+  ): string {
+    if (typeMap.primitive) {
+      return typeMap.primitive;
+    } else if (typeMap.array || typeMap.map) {
+      const prefix = typeMap.array ? '[]' : 'map[string]';
       const innerName =
-        new GoTypeRef(this.root, this.reference.mapOfType).scopedTypeName(
+        this.scopedTypeName(
+          (typeMap.array ?? typeMap.map!).typeMap,
```

Comment:
Optional (I'm even starting to doubt myself here) - I kinda find moving the bang after the parentheses makes it better signal "one of these is defined"... I don't know...

```suggestion
          (typeMap.array ?? typeMap.map)!.typeMap,
```